eclair-core/src/main/scala/fr/acinq/eclair/channel/fsm/Channel.scala

```diff
@@ -1588,7 +1585,47 @@ class Channel(val nodeParams: NodeParams, val wallet: OnChainChannelFunder with
     // peer doesn't cancel the timer
     case Event(TickChannelOpenTimeout, _) => stay()
 
-    // we declare WatchFundingSpentTriggered handlers here because they apply to variants of each state in OFFLINE/SYNCING
+    case Event(w: WatchPublishedTriggered, d: PersistentChannelData) =>
+      log.info(s"zero-conf funding txid=${w.tx.txid} has been published")
+      val fundingStatus = LocalFundingStatus.ZeroconfPublishedFundingTx(w.tx)
+      val metaCommitments1 = d.metaCommitments.updateLocalFundingStatus(w.tx.txid, fundingStatus)
+      val d1 = d match {
+        case d: DATA_WAIT_FOR_FUNDING_CONFIRMED => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED =>
+          val realScidStatus = RealScidStatus.Unknown
+          val shortIds = createShortIds(d.channelId, realScidStatus)
+          // NB: we discard remote's stashed channel_ready, they will send it back at reconnection
+          DATA_WAIT_FOR_DUAL_FUNDING_READY(metaCommitments1, shortIds)
+        case d: DATA_WAIT_FOR_CHANNEL_READY => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_WAIT_FOR_DUAL_FUNDING_READY => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_NORMAL => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_SHUTDOWN => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_NEGOTIATING => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_WAIT_FOR_REMOTE_PUBLISH_FUTURE_COMMITMENT => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_CLOSING => d // there is a dedicated handler in CLOSING state
+      }
+      watchFundingConfirmed(w.tx.txid, Some(nodeParams.channelConf.minDepthBlocks))
+      stay() using d1 storing()
+
+    case Event(w: WatchFundingConfirmedTriggered, d: PersistentChannelData) =>
+      log.info(s"funding txid=${w.tx.txid} has been confirmed")
+      val metaCommitments1 = acceptFundingTxConfirmed(w, d)
+      val d1 = d match {
+        case d: DATA_WAIT_FOR_FUNDING_CONFIRMED => d.copy(metaCommitments = metaCommitments1)
+        case d: DATA_WAIT_FOR_DUAL_FUNDING_CONFIRMED =>
+          val realScidStatus = RealScidStatus.Temporary(RealShortChannelId(w.blockHeight, w.txIndex, d.metaCommitments.latest.commitInput.outPoint.index.toInt))
```

Comment:
I believe that using `d.metaCommitments.latest` here is incorrect, it's not necessarily the latest funding transaction that just confirmed?

Reply:
I still don't understand that, I must be missing something. I'm thinking of the following scenario:

- we initiate dual funding and create fundingTx1
- we RBF, creating fundingTx2
- we go offline
- fundingTx1 confirms

The temporary `RealShortChannelId` that the code will currently create will be invalid: it will use the funding output index in `fundingTx2` instead of using the funding output index in `fundingTx1` (`d.metaCommitments.latest.commitInput.outPoint.index.toInt`). I believe we should use `metaCommitments1.latest.commitInput.outPoint.index.toInt`.

Reply:
Indeed! Fixed in bd65f596d.